mitodl/bootcamp-ecommerce
# Refactor hubspot deal/line syncing

### What are the relevant tickets?
Closes #545 
Closes #542 

#### What's this PR do?
- Refactors hubspot syncing code to base serialized messages on `BootcampApplication` instead of `PersonalPrice`
- Adds a signal to trigger the hubspot deal sync whenever a `BootcampApplication` is created or saved.
- Adds a signal to trigger the hubspot deal sync whenever an `ApplicationStepSubmission` is created (indicating that one of the application's required steps has been submitted).
- Sets the deal `application_stage` value on hubspot to `BootcampApplication.state` if it is anything other than `AWAITING_USER_SUBMISSIONS `.  Otherwise, sets it to either `AWAITING_VIDEO_INTERVIEW` or `AWAITING_QUIZ_SUBMISSION` depending on the first application step that doesn't have a submission yet.

#### How should this be manually tested?
- Populate cybersource and hubspot env values.
- If you've never done so before, run `manage.py sync_hubspot_data`.  It should complete without errors.  If you have any existing applications, they should show up in hubspot within a few minutes.  
- Manually create an application or two via django admin or shell and save.  The applications should show up under hubspot deals within a few minutes, with an associated contact and product.
- Update the state of an application and save.  The change should be reflected in hubspot.
- Associate a fulfilled order with an application via django admin or shell
- Run `manage.py sync_hubspot_data` again, it should complete without errors and the hubspot deal 'Total price paid' should reflect the order amount.

#### Any background context you want to provide?
Currently, when paying for a run, there is no code that associates the order with the application or ensures that the user has an application before making a payment.  My assumption is that this will be handled later as a separate issue.
